File: 134-gas-station/134-gas-station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        
        int idx = 0;
        int sum_diff = accumulate(gas.begin(),gas.end(),0) - \
            accumulate(cost.begin(),cost.end(),0);
        
        
        if(sum_diff < 0) return -1;
        
        int fuel = 0;
        for(int j=0;j<n;j++){
            fuel += gas[j]-cost[j];
            if(fuel < 0){
                idx = j+1;
                fuel = 0;
            }
            
            
        }
        
        return idx;
    }
};

File: 134-gas-station/134-gas-station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        long total_cost = 0;
        long total_gas = 0;
        int n = gas.length;
        
        for(int i=0;i<n;i++){
            total_cost += cost[i];
        }
        for(int i=0;i<n;i++){
            total_gas += gas[i];
        }
        if(total_gas<total_cost)
            return -1;
        long total = 0;
        int ans_index = 0;
        for(int i=0;i<n;i++){
            total += gas[i];
            total -= cost[i];
            if(total<0)
            {
                ans_index = i+1;
                total = 0;
            }
        }
        return ans_index;
    }
}
